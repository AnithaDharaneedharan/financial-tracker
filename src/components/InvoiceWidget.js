import React, { useState } from "react";
import InvoiceWidgetRow from "../components/InvoiceWidgetRow";
import { setData } from '../services/localData'
import { Table, Button } from "react-bootstrap";

export default function InvoiceWidget({localStorageData, invoices}) {
  
  const [invoiceState, setInvoice] = useState(invoices);

  const addNewInvoice = (newInvoice) => {
    setInvoice([...invoiceState, newInvoice]);
  };

  const updateInvoice = (invoiceObj) => {
    const data = invoiceState;
    let alreadyInWidget = false;

    data.forEach((invoice) => {
      if (invoice.referenceNumber === invoiceObj.referenceNumber) {
        alreadyInWidget = true;
      }
    });
    if (!alreadyInWidget) {
      const localInvoice = [...localStorageData.invoices];
      localInvoice.push(invoiceObj);
      setData("data", {
        ...localStorageData,
        invoices: localInvoice,
      });
      setInvoice([...invoiceState, invoiceObj]);
      return;
    }
    const existingInvoices = [...invoiceState];
    const updatedIndex = existingInvoices.findIndex(
      (existingInvoice) =>
        existingInvoice.referenceNumber === invoiceObj.referenceNumber
    );
    if (updatedIndex < 0) return;
    existingInvoices[updatedIndex] = invoiceObj;
    setData("data", { ...localStorageData, ...{ invoices: existingInvoices } });
    setInvoice(existingInvoices);
  };

  const removeInvoice = (invoiceObj) => {
    const data = invoiceState;
    const newData = data.filter(
      (invoice) => invoice.referenceNumber !== invoiceObj.referenceNumber
    );


    setData("data", { ...localStorageData, ...{ invoices: newData } });
    setInvoice(newData);
  };

  return (
    <div className="heading">
      Add or Modify Invoice Details
      <Table striped bordered hover variant="dark">
        <thead>
          <tr>
            <th>Invoice Reference Number</th>
            <th>Customer Name</th>
            <th>Creation Date</th>
            <th>Amount</th>
            <th>Payment Status</th>
            <th>Action</th>
          </tr>
        </thead>
        <tbody>
          {invoiceState.map((invoice, index) => (
            <InvoiceWidgetRow
              key={index}
              invoice={invoice}
              onSaveEvent={updateInvoice}
              onRemove={removeInvoice}
              transactionDetails={localStorageData.transactionDetails}
            />
          ))}
        </tbody>
      </Table>
      <Button data-testid="add-new-invoice" onClick={() => addNewInvoice({ })}>
        Add New Invoice
      </Button>
    </div>
  );
}
