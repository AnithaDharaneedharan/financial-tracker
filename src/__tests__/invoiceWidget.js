import '@testing-library/jest-dom';
import React from 'react';
import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import InvoiceWidget from '../components/InvoiceWidget';
import mockbankAccDetails from '../services/bankAccDetails.json';

test('should render snapshot', () => {
  const { queryByTestId, asFragment } = render(
    <InvoiceWidget
      localStorageData={mockbankAccDetails}
      invoices={mockbankAccDetails.invoices}
    />
  );
  expect(queryByTestId('add-new-invoice')).toBeTruthy();
  expect(asFragment()).toMatchSnapshot();
});

test('should check if Add New Invoice button is present ', () => {
  const { queryByTestId } = render(
    <InvoiceWidget
      localStorageData={mockbankAccDetails}
      invoices={mockbankAccDetails.invoices}
    />
  );
  expect(queryByTestId('add-new-invoice')).toBeTruthy();
});
