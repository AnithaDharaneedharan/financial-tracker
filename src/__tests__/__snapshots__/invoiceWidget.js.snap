// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="heading"
  >
    Add or Modify Invoice Details
    <table
      class="table table-dark table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Invoice Reference Number
          </th>
          <th>
            Customer Name
          </th>
          <th>
            Creation Date
          </th>
          <th>
            Amount
          </th>
          <th>
            Payment Status
          </th>
          <th>
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <input
              disabled=""
              type="text"
              value="NEXT123"
            />
          </td>
          <td>
            <input
              disabled=""
              type="text"
              value="NEXT"
            />
          </td>
          <td>
            <input
              disabled=""
              type="text"
              value="2020-08-05"
            />
          </td>
          <td>
            <input
              disabled=""
              type="text"
              value="6000"
            />
          </td>
          <td>
            <span
              class="paid"
            >
              PAID
            </span>
          </td>
          <td>
            <span>
              N/A
            </span>
          </td>
        </tr>
        <tr>
          <td>
            <input
              type="text"
              value="WAITROSE123"
            />
          </td>
          <td>
            <input
              type="text"
              value="WAITROSE"
            />
          </td>
          <td>
            <input
              type="text"
              value="2020-08-08"
            />
          </td>
          <td>
            <input
              type="text"
              value="600"
            />
          </td>
          <td>
            <span
              class="unpaid"
            >
              NOT PAID
            </span>
          </td>
          <td>
            <span
              class="action"
            >
              <button
                class="btn btn-secondary"
                type="button"
              >
                Save
              </button>
              <button
                class="btn btn-secondary"
                type="button"
              >
                Remove
              </button>
            </span>
          </td>
        </tr>
        <tr>
          <td>
            <input
              type="text"
              value="UBER123"
            />
          </td>
          <td>
            <input
              type="text"
              value="UBER"
            />
          </td>
          <td>
            <input
              type="text"
              value="2020-08-08"
            />
          </td>
          <td>
            <input
              type="text"
              value="-40"
            />
          </td>
          <td>
            <span
              class="unpaid"
            >
              NOT PAID
            </span>
          </td>
          <td>
            <span
              class="action"
            >
              <button
                class="btn btn-secondary"
                type="button"
              >
                Save
              </button>
              <button
                class="btn btn-secondary"
                type="button"
              >
                Remove
              </button>
            </span>
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="btn btn-primary"
      data-testid="add-new-invoice"
      type="button"
    >
      Add New Invoice
    </button>
  </div>
</DocumentFragment>
`;
